### DISCUSSION TITLE
Using the same VM instance for isolated contexts

### DISCUSSION THREAD

**Discussion by RGBCube** [mlua-rs/mlua#331] [Q&A] [answered]
So, I am making a service where each user provides some Luau code, and I run each one seperately, concurrently.

Is there a way to utilize a single Lua struct to do this instead of doing `DashMap<Key, Lua>` and having a single Lua object?

Also, I do not want the code of different users to affect each other. I want it all to be isolated.

**Comment by khvzak** [comment, marked as answer]
You could try setting [sandbox](https://docs.rs/mlua/latest/mlua/struct.Lua.html#method.sandbox) mode for Luau and executing each chunk in isolated environment.

There are few options:
1) Setting sandbox before executing chunk and then destroyng it (exactly as in the doc example)
2) Setting sandbox and then utilizing per-`Thread` sandbox [see this doc](https://docs.rs/mlua/latest/src/mlua/thread.rs.html#333). Sorry it's a hidden functionality but should work!
3) Setting sandbox and then setting isolated proxy [environment](https://docs.rs/mlua/latest/mlua/struct.Chunk.html#method.set_environment) to a chunk.

You can even make environment [readonly](https://docs.rs/mlua/latest/mlua/struct.Table.html#method.set_readonly) and prevent any writes to globals.

Personally I would try №3

**Reply by RGBCube** [reply]
Thanks! I'll try these soon
